File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        int opening = 0, closing = 0, len = s.length();
        StringBuilder sb = new StringBuilder();
        for(int i=0;i<len;i++){
            char ch = s.charAt(i);
            if(ch == ')'){
                if(opening > 0){
                    sb.append(ch);
                    opening--;
                }else closing++;
            }else if(ch == '('){
                opening++;
                sb.append(ch);
            }else{
                sb.append(ch);
            }
        }
        if(opening > 0){
            for(int i=sb.length()-1;i>=0 && opening > 0;i--){
                if(sb.charAt(i) =='('){
                    opening--;
                    sb.deleteCharAt(i);
                }
            }
        }
        return sb.toString();
    }
}

File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.cs
using System;
using System.Text;

public class Solution {
    public string MinRemoveToMakeValid(string s) {
        StringBuilder str = new StringBuilder();
        int opening = 0, closing = 0;
        for(int i=0;i<s.Length;i++){
            char ch = s[i];
            if(ch == ')'){
                if(opening > 0){
                    opening--;
                    str.Append(ch);
                }else closing++;
            }else if(ch == '('){
                str.Append(ch);
                opening++;
            }else{
                str.Append(ch);
            }
        }
        if(opening > 0){
            for(int i=str.Length-1;i>=0 && opening > 0;i--){
                if(str[i] == '('){
                    opening--;
                    str.Remove(i,1);
                }
            }
        }
        return str.ToString();
    }
}